dotnet/src/SemanticKernel.Abstractions/AI/FunctionChoiceBehaviors/FunctionChoiceBehavior.cs

```diff
@@ -0,0 +1,72 @@
+[JsonDerivedType(typeof(NoneFunctionChoiceBehavior), typeDiscriminator: NoneFunctionChoiceBehavior.TypeDiscriminator)]
+[Experimental("SKEXP0001")]
+public abstract class FunctionChoiceBehavior
+{
+    /// <summary>
+    /// Creates a new instance of the <see cref="FunctionChoiceBehavior"/> class.
+    /// </summary>
+    internal FunctionChoiceBehavior()
+    {
+    }
+
+    /// <summary>
+    /// Gets an instance of the <see cref="FunctionChoiceBehavior"/> that provides either all of the <see cref="Kernel"/>'s plugins' function information to the model or a specified subset.
+    /// This behavior allows the model to decide whether to call the functions and, if so, which ones to call.
+    /// </summary>
+    /// <param name="autoInvoke">Indicates whether the functions should be automatically invoked by the AI service/connector.</param>
+    /// <param name="functions">The subset of the <see cref="Kernel"/>'s plugins' functions to provide to the model.
+    /// If null or empty, all <see cref="Kernel"/>'s plugins' functions are provided to the model.</param>
+    /// <returns>An instance of one of the <see cref="FunctionChoiceBehavior"/>.</returns>
+    public static FunctionChoiceBehavior Auto(bool autoInvoke = true, IEnumerable<KernelFunction>? functions = null)
+    {
+        return new AutoFunctionChoiceBehavior(autoInvoke, functions);
+    }
+
+    /// <summary>
+    /// Gets an instance of the <see cref="FunctionChoiceBehavior"/> that provides either all of the <see cref="Kernel"/>'s plugins' function information to the model or a specified subset.
+    /// This behavior forces the model to always call one or more functions. The model will then select which function(s) to call.
+    /// </summary>
+    /// <param name="autoInvoke">Indicates whether the functions should be automatically invoked by the AI service/connector.</param>
+    /// <param name="functions">The subset of the <see cref="Kernel"/>'s plugins' functions to provide to the model.
+    /// If null or empty, all <see cref="Kernel"/>'s plugins' functions are provided to the model.</param>
+    /// <returns>An instance of one of the <see cref="FunctionChoiceBehavior"/>.</returns>
+    public static FunctionChoiceBehavior Required(bool autoInvoke = true, IEnumerable<KernelFunction>? functions = null)
+    {
+        return new RequiredFunctionChoiceBehavior(autoInvoke, functions);
+    }
+
+    /// <summary>
+    /// Gets an instance of the <see cref="FunctionChoiceBehavior"/> that provides either all of the <see cref="Kernel"/>'s plugins' function information to the model or a specified subset.
+    /// This behavior forces the model to not call any functions and only generate a user-facing message.
+    /// </summary>
+    /// <param name="functions">The subset of the <see cref="Kernel"/>'s plugins' functions to provide to the model.
+    /// If null or empty, all <see cref="Kernel"/>'s plugins' functions are provided to the model.</param>
```

Comment:
If empty I would throw `ArgumentException`.

```suggestion
    /// If not provided, defaults to all plugins functions currently added to the <see cref="Kernel"/> instance.</param>
```

Reply:
Good point, we can do it. Alternatively, we may treat empty array as a way to turn off function calling:
- The `FunctionChoiceBehavior.Auto(functions: listOfFuncitons)` call would advertise only functions from listOfFuncitons.
- The `FunctionChoiceBehavior.Auto(functions: null)` call would advertise all functions registered on kernel.
- The `FunctionChoiceBehavior.Auto(functions: [])` call would advertise no functions and can be used to disable function calling.

Reply:
The latter suggestion sounds reasonable.

Reply:
> The latter suggestion sounds reasonable, i.e. empty means no functions

There are two ways to handle an empty array. The first way is to disable function calling. The second way is to set `options.ToolChoice` to auto and assign an empty array to the `options.Tools` property. In the second case, AzureSDK will throw an exception stating that no tools were provided for the auto choice. ToolCallBehavior uses the first way, and I more inclined to follow it.

> I'd also be happy not having a list here at all, and just using what's in the kernel, as is the case today.

Do you mean to remove the functions parameter of the Auto static method? Today, tool call behavior allows to supply functions via the `ToolCallBehavior.EnableFunctions(IEnumerable<OpenAIFunction> functions, bool autoInvoke = false)` method.

Reply:
> ToolCallBehavior uses the first way, and I more inclined to follow it

Yes, I don't think we should throw.

> Today, tool call behavior allows to supply functions via the ToolCallBehavior.EnableFunctions(IEnumerable<OpenAIFunction> functions, bool autoInvoke = false) method.

Yeah.  Ok.